warehouse/tuf/services.py

```diff
@@ -0,0 +1,398 @@
+
+        hash_bins = self._get_hash_bins()
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # Delegate first targets -> BINS
+        delegate_roles_payload = dict()
+        delegate_roles_payload["targets"] = list()
+        delegate_roles_payload["targets"].append(
+            RolesPayload(
+                expiration=self._set_expiration_for_role(Role.BINS.value),
+                threshold=self._request.registry.settings[
+                    f"tuf.{Role.BINS.value}.threshold"
+                ],
+                keys=self._key_storage_backend.get(Role.BINS.value, "private"),
+                delegation_role=Role.BINS.value,
+                paths=["*", "*/*/*/*"],
+            )
+        )
+
+        metadata_repository.delegate_targets_roles(
+            delegate_roles_payload,
+            self._set_expiration_for_role(Role.TIMESTAMP.value),
+            self._set_expiration_for_role(Role.SNAPSHOT.value),
+        )
+
+        # Delegates all BINS -> BIN_N (Hash bin prefixes)
+        delegate_roles_payload = dict()
+        delegate_roles_payload[Role.BINS.value] = list()
+        for bin_n_name, bin_n_hash_prefixes in hash_bins.generate():
+            delegate_roles_payload[Role.BINS.value].append(
+                RolesPayload(
+                    expiration=self._set_expiration_for_role(Role.BIN_N.value),
+                    threshold=self._request.registry.settings[
+                        f"tuf.{Role.BIN_N.value}.threshold"
+                    ],
+                    keys=self._key_storage_backend.get(Role.BIN_N.value, "private"),
+                    delegation_role=bin_n_name,
+                    path_hash_prefixes=bin_n_hash_prefixes,
+                )
+            )
+
+        metadata_repository.delegate_targets_roles(
+            delegate_roles_payload,
+            self._set_expiration_for_role(Role.TIMESTAMP.value),
+            self._set_expiration_for_role(Role.SNAPSHOT.value),
+        )
+
+    def bump_snapshot(self):
+        """
+        Bump Snapshot Role Metadata
+        """
+        # Bumping the Snapshot role involves the following steps:
+        # 1. Initiate Metadata Repository.
+        # 2. Load the Snapshot Role.
+        # 3. Bump Snapshot role (and write to the Storage).
+        # 4. Bump Timestamp role using the new Snapshot Metadata.
+        # 1. Metadata Repository.
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # 2. Snapshot role metadata.
+        snapshot_metadata = metadata_repository.load_role(Role.SNAPSHOT.value)
+
+        # 3. Bump Snapshot role metadata.
+        snapshot_metadata = metadata_repository.snapshot_bump_version(
+            snapshot_expires=self._set_expiration_for_role(
+                self._request, Role.SNAPSHOT.value
+            ),
+            snapshot_metadata=snapshot_metadata,
```

Comment:
Couldn't we omit the optional `snapshot_metadata` here and let `snapshot_bump_version` call `load_role`? This would also be consistent with the usage of `timestamp_bump_version` below, which always calls `load_role` by itself.

Reply:
Yes. I see the point.